### PR TITLE
Add support for legends

### LINKED ISSUE #29: Add ability to show axes labels
I can think of three ways to activate axes labels:

1. Using the column names of the dataframe (this would only if the user passing in the dataframe via `data`)
    ```py
    scatter.axes(True, labels=True)
    ```

2. Using custom labels
    ```py
    scatter.axes(True, labels=['x', 'y'])
    scatter.axes(True, labels=dict(x='x', y='y'))
    ```

### PR DESCRIPTION
**Pull request by flekschas**
This PR addresses #29 and adds support for legends.

Screen Shot 2022-07-20 at 1 29 38 PM

The legend documents the point color, opacity, and size encoding at the moment. For categorical data, all categories are visualized. For continuous data, five points on the scale are visualized.

### API

```py
jscatter.Scatter(..., legend=True, legend_position='top-left')

# or

scatter = jscatter.Scatter(...)
scatter.legend(True, position='top-left')
```

### To-Dos

- [x] Show point connection color, opacity, and size encoding as well
- [x] Add dark mode support

### Issue

Closes #29

### CODE REVIEW

File: jscatter/jscatter.py [hunk 1]
```diff
@@ -62,22 +65,22 @@ def order_map(map, order):
...
     # same way. It create a texture and based on the point value
     # accesses a color. The point values are first normalized to [0, 1]
     # given the value range and then mapped to the range of colors.
     # E.g., if you have data values in [0, 10] and 5 colors,
     # 6.7 maps to the color with index floor(5 * 6.7/10) = 3. The same
     # principle is applied to categorical data! This means we need to
-    # ensure that the number of colors is the same as the number of
-    # categories otherwise weird mapping glitches can appear.
+    # ensure that the number of map values is the same as the number of
+    # categories otherwise weird mapping glitches can happe .
```
Comment (manzt) [inline jscatter/jscatter.py:79]:
```suggestion
    # categories otherwise weird mapping glitches can happen.
```

File: jscatter/jscatter.py [hunk 2]
```diff
@@ -2497,7 +2572,64 @@ def axes(
             labels = self._axes_labels,
         )
 
-        return self._mouse_mode
+    def legend(
+        self,
+        legend: Optional[Union[bool, Undefined]] = UNDEF,
+        position: Optional[Union[Position, Undefined]] = UNDEF,
```
Comment (manzt) [inline jscatter/jscatter.py:2578]: these signatures could be simplified as `Union[SomeType, Undefined, None]`. See https://fastapi.tiangolo.com/python-types/#using-union-or-optional. Fairly subjective point, but something I've adopted in projects recently.

EDIT: Ah... I see the style in the rest of this module uses `Optional` maybe something to think about for a follow-up PR but not necessary.

**Review by manzt — APPROVED** [review]
Really nice work, @flekschas! Just some minor comments.

**Comment by manzt** [comment]
my only comment is that the labels seem a little small (and I noticed they were hard-coded). (I'm using FireFox).

image

File: jscatter/encodings.py [hunk 3]
```diff
@@ -72,6 +84,13 @@ def delete(self, encoding):
                 break
 
 
+class VisualEncoding():
+    def __init__(self, channel, data, legend=None):
+        self.channel = channel
+        self.data = data
+        self.legend = legend
```
[the remark this answers is not in the document]
Comment (flekschas) [reply jscatter/encodings.py:91]: Oh sweet! That's a great idea. I didn't even know of dataclasses!

File: jscatter/jscatter.py (hunk 2 above)
Comment (flekschas) [reply jscatter/jscatter.py:2578]: Ha! That's interesting. I think I'll stick to `Optional` for now but I guess `Union[bool, Undefined]` would be sufficient as well. (`None` is not valid in this case. I remember we talked about this but in order to distinguish (elsewhere) between unsetting and ignoring a value I introduced `Undefined`. This way I can check if the user provided no value versus `None`.)

**Comment by flekschas** [comment]
Thanks a lot for the quick review. It's interesting that the legend renders differently on your machine. It seems like the text is smaller but more spaced out compared to when I render it in FF. This might have to do with the base font-size. It might be better to switch to absolute sizes and allow the user to pick one of _small_, _medium_, or _large_.

**Comment by flekschas** [comment]
@manzt Do you think you could give the PR another quick test? The staleness should be fixed. The legend should also be rendered more consistently now and you can control the size via `legend(size='medium')` (It accepts small, medium, or large).

**Review by manzt — APPROVED** [review]
Tried out the lastest changes. Looks great!

**Pull request merged**